sdk/provisioning/Azure.Provisioning/src/BicepDictionaryOfT.cs

```diff
@@ -75,8 +76,34 @@ public static implicit operator BicepDictionary<T>(ProvisioningVariable referenc
     /// <returns>The value.</returns>
     public BicepValue<T> this[string key]
     {
-        get => _values[key];
-        set => _values[key] = value;
+        get
+        {
+            if (!_values.TryGetValue(key, out var existingValue))
+            {
+                return new BicepDictionaryIndexer<T>(_self, key);
+            }
+            // now we have a value
+            var value = (IBicepValue)existingValue;
+            switch (value.Kind)
+            {
+                case BicepValueKind.Unset:
+                    return new BicepDictionaryIndexer<T>(_self, key);
+                case BicepValueKind.Expression:
+                    return new BicepDictionaryIndexer<T>(_self, key, value.Expression!);
+                case BicepValueKind.Literal:
+                    return new BicepDictionaryIndexer<T>(_self, key, (T)value.LiteralValue!);
+                default:
+                    throw new InvalidOperationException($"Unknown {value.Kind}!");
+            }
+        }
+        set
+        {
+            if (_kind == BicepValueKind.Expression || _isOutput)
```

Comment:
what happened before if someone tried to set such a value?

Reply:
they will succeed at setting it.